src/typing/env_js.ml

```diff
@@ -661,24 +771,21 @@ let merge_env =
           (string_of_reason reason)
           name)
       (* global lookups may leave new entries in one or both child envs *)
-      | None, child1, child2 when env0 = [] ->
+      | None, _, _ when env0_ = [] ->
         (* ...in which case we can forget them *)
         ()
-      (* otherwise, non-refinement uneven distributions are asserts. *)
-      (* TODO:
-         Asserting on all uneven distributions is too strict for now.
-         let_env can interact with path-dependent refinement to leave
-         residual entries in child contexts, when catch blocks combine
-         conditionals and early returns.
-         So for now we need to let extra child entries pass, though we
-         should still assert on extra parent entries.
-         Once let/const/LexScope is in place, we can reimplement let_env
-         in terms of those constructs instead, and the full assert can
-         be reinstated.
-       *)
-      | None, _, _ ->
+      (* changeset entry exists only in lex scope *)
+      | None, None, Some e when is_lex scope2 && Entry.is_lex e ->
+        ()
+      | None, Some e, None when is_lex scope1 && Entry.is_lex e ->
         ()
-      | Some _ as orig, child1, child2 ->
+      (* walk through uneven lex scopes *)
```

Comment:
Think I found it & it's a bug in the env copying. Going to try to write a
few tests to suss out the expected behavior, because I think there is a
correlated bug.

On Tue, Sep 8, 2015 at 3:57 PM, Basil Hosmer <notifications@github.com>
wrote:

> In src/typing/env_js.ml
> <https://github.com/facebook/flow/pull/784#discussion_r39000752>:
>
> >          ()
> > -      | Some _ as orig, child1, child2 ->
> > +      (* walk through uneven lex scopes *)
>
> Hmmm, ok. LMK if you don't find an error and I can look upstream and see
> if I can figure out why it would be happening. Seems like it shouldn't be.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/flow/pull/784/files#r39000752>.
>